### ISSUE TITLE
prov/shm: smr_progress_ipc_list() causes 200ns performance regression in 256 rank osu_gather benchmark

### ISSUE THREAD

**Issue by a-szegel** [ofiwg/libfabric#9141]
**Describe the bug**
This commit causes a 200ns performance regression in osu_gather benchmark.
```
commit 6c86459680e00af11a98d43d60547eb2f239b7df (HEAD)
Author: Amir Shehata <shehataa@ornl.gov>
Date:   Thu Nov 24 16:59:33 2022 -0800

    prov/shm: Update SHM to use ROCR
    
    Update the SHM provider to use the ROCR HMEM asynchronous memory
    operations.
    
       . Unify the ipc and sar freestack, since they use the same structure.
       . When progressing an IPC operation, check if the device is ROCR and trigger
         an asynchronous operation.
       . When an asynchronous operation is queued, create an IPC entry and
         add it to the queue of pending operations.
       . During the top level progress loop check the queue of pending asynchronous
         operations and query the state of each one. Generate a complete event
         for finished operations. Since completions happen outside the context
         of libfabric we can't rely on the ep->region->signal flag to be set.
         Always check the pending queue. This shouldn't introduce much of an
         overhead if the queue is empty
       . use the ep->util_ep lock to protect the free stack and the ipc list
         of pending operations.
    
    Signed-off-by: Amir Shehata <shehataa@ornl.gov>
```
I isolated the cause of the regression to be this: https://github.com/ofiwg/libfabric/blob/main/prov/shm/src/smr_progress.c#L1287

This regression is mostly fixed if we move `smr_progress_ipc_list(ep);` into 
```
	if (ofi_atomic_cas_bool32(&ep->region->signal, 1, 0)) {
		if (smr_env.use_dsa_sar)
			smr_dsa_progress(ep);
		smr_progress_resp(ep);
		smr_progress_cmd(ep);
		smr_progress_sar_list(ep);
		smr_progress_ipc_list(ep);
	}
```

There is still an extra 50ns overhead when I compare to commenting `smr_progress_ipc_list()` out completely (which I understand isn't ok b/c it is required for correctness with hmem ipc capable devices).

Is it possible to make polling the ipc_list lighter (and other lists)?
Is it possible to move polling the IPC list inside the signal if statement and maintain correctness?
Is it possible to get rid of this polling at runtime if we do not have any GPU's on the instance?

**To Reproduce**
Run osu_gather on 8 c6gn.16xlarge with 32 processes per node (256 rank total).

**Environment:**
Amazon Linux 2

**Comment by sunkuamzn** [comment]
How about we have a bool `SMR_ASYNC_DEVICE_COPY` which is only set if ROCR is initialized? The `ipc_copy_pend_list` member in `smr_ep` and all related async device copy stuff is only available if `SMR_ASYNC_DEVICE_COPY` is true. If in the future CUDA devices also use this async protocol, then we can set `SMR_ASYNC_DEVICE_COPY` for CUDA too. What do you think @aingerson?

**Comment by aingerson** [comment]
@sunkuamzn That is ok with me.

**Comment by a-szegel** [comment]
Merged into Main, PR to merge into 1.19.x: https://github.com/ofiwg/libfabric/pull/9202
